sysvabi64/sysvabi64.rst

```diff
@@ -1414,30 +1414,57 @@ below.
 GNU C Library IFUNC interface
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The GNU C Library (glibc) ``sys/ifunc.h`` header provides the
-necessary type definitions for defining IFUNC resolvers. The resolver
-is passed at least one, and at most 2 parameters dependent on the
-glibc IFUNC ABI version. A flag is set in the first parameter if a
-second parameter is available. The parameters provide access to the
-HWCAP_ flags.
+The prototype of a GNU indirect function resolver is:
 
 .. code-block:: c
 
-  /* A second argument is passed to the ifunc resolver.  */
-  #define _IFUNC_ARG_HWCAP        (1ULL << 62)
+   ElfW(Addr) ifunc_resolver (uint64_t, const uint64_t *);
 
-  /* Function prototype for an IFUNC resolver.
-  ElfW(Addr) ifunc_resolver (uint64_t, const __ifunc_arg_t *);
+The resolver is passed at least one, and at most 2 parameters dependent
+on the GNU C Library (glibc) IFUNC ABI version. The bits in the first parameter
+will match the ``AT_HWCAP`` settings except for the ``_IFUNC_ARG_HWCAP`` bit
+that has special meaning: if this bit is set, then the second parameter is
+passed to the resolver function.
 
-  /* Second argument to an ifunc resolver.  */
-  struct __ifunc_arg_t
-  {
-    unsigned long _size; /* Size of the struct, so it can grow.  */
-    unsigned long _hwcap;
-    unsigned long _hwcap2;
-  };
+.. code-block:: c
+
+   /* A second parameter is passed to the ifunc resolver.  */
+   #define _IFUNC_ARG_HWCAP (1ULL << 62)
+
+When the second parameter is passed, it will be a const pointer to a buffer
+that allows to access all HWCAP elements. The buffer may contain the following
+fields:
+
+``_size``
+   an ``uint64_t`` value at byte offset 0 containing the size of the buffer
```

Comment:
In the category of English is an awful language, whilie u is a vowel and if we spelled out unsigned integer, `an` would be correct, when it is an abbreviation `an uint64_t` doesn't read well. It seems like `A uint64_t` is more common and reads a lot better.

We could potentially avoid that by just starting with ``uint64_t``

Reply:
or just 'unsigned long', since that's what the corresponding code uses...

Reply:
@enh-google what are your thoughts on using `uint64_t` in the libc header instead of `unsigned long`? We really use those value as 64 bit integers and `unsigned long` is not always 64 bit on some platforms.

Reply:
but this is specifically only for arm64 :-)

(my personal preference in new api is always for the more specific types, but i suspect what happened here was that they copied the type from getauxval(), which is where these values come from. so left as `unsigned long` it's potentially reusable for other architectures too, where `uint64_t` would make it LP64-only. i have no idea whether glibc has any plans to use this for other architectures. android only supports arm64, riscv64, and x86-64, and they're all sufficiently different from one another that the sharing wouldn't really make sense.)

Reply:
I've fixed the cosmetic suggestion and kept `uint64_t` because I think it's more correct to describe it this way. We use bits above 32 in both `hwcap` and `hwcap2` already. I'll keep `unsigned long` in the header though.